File: problem_10/prime_summation_test.rb
require 'minitest/autorun'
require 'minitest/pride'

class PrimeSummationTest < Minitest::Test
  def prime_summation(max)
    max = max - 1 if max.even?
    sum = max >= 2 ? 2 : 0

    (3..max).step(2) do |num|
      prime = true
      (2..(num ** 0.5 + 1)).each do |n|
        break prime = false if num % n == 0
      end
      sum += num if prime
    end

    sum
  end

  #----------------- Tests -----------------#

  def test_prime_summation
    assert_equal 17, prime_summation(10)
    assert_equal 76127, prime_summation(1000)
    assert_equal 142913828922, prime_summation(2000000)
  end
end

File: problem_10/prime-summation.test.js
function primeSummation(maxNum) {
  let sum = 0;
  for (n = 2; n < maxNum; n++) {
    if (prime(n)) {
      sum += n;
    }
  }

  return sum;
}

function prime(num) {
  if (num < 2) {
    return false;
  }

  let sqrtMax = Math.pow(num, 0.5);
  for (i = 2; i <= sqrtMax; i++) {
    if (num % i === 0) {
      return false;
    }
  }
  return true;
}

//----------------- Tests -----------------//

test("primeSummation", () => {
  expect(primeSummation(10)).toBe(17);
  expect(primeSummation(1000)).toBe(76127);
  expect(primeSummation(2000000)).toBe(142913828922);
});

test("prime", () => {
  expect(prime(0)).toBe(false);
  expect(prime(1)).toBe(false);
  expect(prime(2)).toBe(true);
  expect(prime(4)).toBe(false);
  expect(prime(5)).toBe(true);
  expect(prime(29)).toBe(true);
  expect(prime(99)).toBe(false);
});
